### PR TITLE
Email unique constraint by email and verified: true

### PR DESCRIPTION
**Pull request by techgaun**
@ericmj what do you think of this change for making email constraint based on email and verified true?

### CODE REVIEW

File: mix.exs [hunk 1]
```diff
@@ -77,7 +77,7 @@ defmodule HexWeb.Mixfile do
   defp aliases do
     ["compile.gpb": &compile_gpb/1,
      "ecto.setup": ["ecto.create", "ecto.migrate", "run priv/repo/seeds.exs"],
-     "ecto.reset": ["ecto.drop", "ecto.create", "ecto.migrate"],
+     "ecto.reset": ["ecto.drop", "ecto.setup"],
```
Comment (ericmj) [inline mix.exs]: I don't want the seed data to run here. `ecto.reset` should put the database in a clean state.

File: web/models/email.ex [hunk 2]
```diff
@@ -41,6 +43,7 @@ defmodule HexWeb.Email do
 
   def verify(email) do
     change(email, %{verified: true, verification_key: nil})
+    |> unique_constraint(:email, name: "emails_email_key", message: "Email already verified by another user")
```
Comment (ericmj) [inline web/models/email.ex]: I don't think we need to mention "verified" here or above. It's enough to say "email already in use".

File: web/models/email.ex [hunk 3]
```diff
@@ -54,4 +57,15 @@ defmodule HexWeb.Email do
   def order_emails(emails) do
     Enum.sort_by(emails, &[not &1.primary, not &1.public, not &1.verified, -&1.id])
   end
+
+  defp validate_verified_email_exists(changeset, field, opts) do
```
Comment (ericmj) [inline web/models/email.ex]: Please move this to `validations.ex` we should avoid accessing the repo inside model modules.

File: web/models/user.ex [hunk 4]
```diff
@@ -110,7 +110,7 @@ defmodule HexWeb.User do
     # Makes assumptions about how Ecto choses variable names
     from u in HexWeb.User,
       where: u.username == ^username_or_email or
-             ^username_or_email in fragment("SELECT emails.email FROM emails WHERE emails.user_id = u0.id"),
+             ^username_or_email in fragment("SELECT emails.email FROM emails WHERE emails.user_id = u0.id and emails.verified"),
```
Comment (ericmj) [inline web/models/user.ex]: Nice catch!

File: priv/repo/migrations/20161030220220_modify_emails_unique_constraint.exs [hunk 5]
```diff
@@ -0,0 +1,15 @@
+defmodule HexWeb.Repo.Migrations.ModifyEmailsUniqueConstraint do
+  use Ecto.Migration
+
+  def up do
+    execute ~s{ALTER TABLE emails DROP CONSTRAINT emails_email_key}
+    execute ~s{CREATE UNIQUE INDEX emails_email_key ON emails (email) WHERE verified = 'true'}
+    execute ~s{CREATE UNIQUE INDEX emails_email_user_key ON emails (email, user_id)}
```
Comment (ericmj) [inline priv/repo/migrations/20161030220220_modify_emails_unique_constraint.exs]: Good idea!

**Comment by techgaun** [comment]
@ericmj sorry about the issues. Updated as per your feedback. Could you please check now?

**Pull request merged**